File: android.sockets/android.Sockets/ClientHelper.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Threading;

namespace android.Sockets
{
	public static class ClientHelper
    {
		public static bool useHeader = false;
		public static bool serverCheck = false;
		public static void InitialiseCommunication(Socket socket, bool isUseHeader,bool isServerCheck = false)
		{
			useHeader = isUseHeader;
			var b = BitConverter.GetBytes(useHeader ? 1 : 2);
            Array.Reverse(b);
            socket.Send(b);

			serverCheck = isServerCheck;
			if (useHeader)
			{
				b = BitConverter.GetBytes(serverCheck ? 1 : 2);
                Array.Reverse(b);
                socket.Send(b);
			}
		}
		public static void sendFileToServer(byte[] dataBytes, String fileName, Socket server){
			if (!useHeader)
				sendFileToServer1(dataBytes, fileName, server);
			else
				sendFileToServer2(dataBytes, fileName, server);
		}
		private static void sendFileToServer1(byte[] dataBytes, String fileName, Socket server){
            String message = "sending";
            byte[] byteMessage = Encoding.UTF8.GetBytes(message);
            int length = byteMessage.Length;
            byte[] b = BitConverter.GetBytes(length);
			Array.Reverse(b);
			server.Send(b);
            server.Send(byteMessage);

            b = new byte[4];
			server.Receive(b);
            Array.Reverse(b);
            int ok = BitConverter.ToInt32(b);
            
			byteMessage = Encoding.UTF8.GetBytes(fileName);
            length = byteMessage.Length;
			b = BitConverter.GetBytes(length);
            Array.Reverse(b);
            server.Send(b);
            server.Send(byteMessage);
            
			var byteStream = new MemoryStream(dataBytes);
			length = dataBytes.Length;

			int bufferSize;
			if (length > 524288)
				bufferSize = 524288;
            else
                bufferSize = length;
 
			b = BitConverter.GetBytes(length);
            Array.Reverse(b);
            server.Send(b);

			byte[] buffer = new byte[bufferSize];

			b = new byte[4];
			server.Receive(b);
            Array.Reverse(b);
            ok = BitConverter.ToInt32(b);
			int count;
			while ((count = byteStream.Read(buffer,0,bufferSize)) > 0)
			{
				server.Send(buffer, count, SocketFlags.None);
            }

			b = new byte[4];
			server.Receive(b);
            Array.Reverse(b);
            length = BitConverter.ToInt32(b);
			byteMessage = new byte[length];
			server.Receive(byteMessage);
			String serverMessage = Encoding.UTF8.GetString(byteMessage);
			if (!serverMessage.Equals("recieved"))
				throw new SocketException();

			return ;
		}

		private static void sendFileToServer2(byte[] dataBytes, String fileName, Socket server)
        {
           

            var byteStream = new MemoryStream(dataBytes);
            int length = dataBytes.Length;


			Header header = new Header(SocketMessageHelper.Sending, fileName, length);
			server.Send(header.getByteHeader(), 55, SocketFlags.None);

            int bufferSize;
			if (length > 524288)
				bufferSize = 524288;
            else
                bufferSize = length;

            byte[] buffer = new byte[bufferSize];

            int count;
            while ((count = byteStream.Read(buffer, 0, bufferSize)) > 0)
            {
				server.Send(buffer, count, SocketFlags.None);
            }

			if (serverCheck)
            {
				byte[] b = new byte[1];
				server.Receive(b, 1, SocketFlags.None);
                if (b[0] != 1)
                    throw new IOException("Server not ok!");
            }

            return;
        }

		public class FileTestData
        {
            public String name;
            public byte[] data;

            public FileTestData() { }

            public FileTestData(String name, byte[] data)
            {
                this.name = name;
                this.data = data;
            }
        }

		public static FileTestData reachFileFromServer(Socket server, String fileName){
			if (!useHeader)
				return reachFileFromServer1(server, fileName);
			else
				return reachFileFromServer2(server, fileName);
		}

		private static FileTestData reachFileFromServer1(Socket server, String fileName){  
			
            String message = "requesting";
            byte[] byteMessage = Encoding.UTF8.GetBytes(message);
            int length = byteMessage.Length;
			byte[] b = BitConverter.GetBytes(length);
            Array.Reverse(b);
            server.Send(b);
            server.Send(byteMessage);

            b = new byte[4];
			server.Receive(b);
            Array.Reverse(b);
			int ok = BitConverter.ToInt32(b);

            byteMessage = Encoding.UTF8.GetBytes(fileName);
            length = byteMessage.Length;
			b = BitConverter.GetBytes(length);
            Array.Reverse(b);
            server.Send(b);
            server.Send(byteMessage);

			b = new byte[4];
			server.Receive(b);
            Array.Reverse(b);
            length = BitConverter.ToInt32(b);
            byteMessage = new byte[length];
            server.Receive(byteMessage);
            String serverMessage = Encoding.UTF8.GetString(byteMessage);

            b = new byte[4];
			server.Receive(b);
            Array.Reverse(b);
            length = BitConverter.ToInt32(b);

            int bufferSize;
			if (length > 524288)
				bufferSize = 524288;
            else
                bufferSize = length;

            b = new byte[4];
			server.Receive(b);
            Array.Reverse(b);
			ok = BitConverter.ToInt32(b);
			Thread.Sleep(1);
			var remoteEndPoint = server.RemoteEndPoint as IPEndPoint;
            remoteEndPoint.Port += 1;
            byte[] buffer = new byte[bufferSize];
            byte[] bytes = new byte[length];
			MemoryStream dataBuffer = new MemoryStream();
			var count = length;
			int dataRead;
			while ((dataRead = server.Receive(buffer, bufferSize, SocketFlags.None)) != -1 )
            {
				count -= dataRead;
				dataBuffer.Write(buffer, 0, dataRead);
                if (count <= 0)
                    break;
				if (count <= bufferSize)
					bufferSize = count;
            }

			Array.Copy(dataBuffer.ToArray(), bytes, length);
			dataBuffer.Close();

			message = "recieved";
			byteMessage = Encoding.UTF8.GetBytes(message);
            length = byteMessage.Length;
			b = BitConverter.GetBytes(length);
            Array.Reverse(b);
            server.Send(b);
            server.Send(byteMessage);

			return new FileTestData(fileName, bytes);
		}

		public static FileTestData reachFileFromServer2(Socket server, String fileName)
        {

            Header header = new Header(SocketMessageHelper.Requesting, fileName, 0);
			server.Send(header.getByteHeader(), 55, SocketFlags.None);

            byte []b = new byte[4];
			server.Receive(b, 4, SocketFlags.None);
            Array.Reverse(b);
            int length = BitConverter.ToInt32(b);

            int bufferSize;
			if (length > 524288)
				bufferSize = 524288;
            else
                bufferSize = length;

            byte[] buffer = new byte[bufferSize];
            byte[] bytes = new byte[length];
            MemoryStream dataBuffer = new MemoryStream();
			var count = length;
            int dataRead;
            while ((dataRead = server.Receive(buffer, bufferSize, SocketFlags.None)) != -1)
            {
                count -= dataRead;
                dataBuffer.Write(buffer, 0, dataRead);
                if (count <= 0)
                    break;
                if (count <= bufferSize)
                    bufferSize = count;
            }

            Array.Copy(dataBuffer.ToArray(), bytes, length);
            dataBuffer.Close();

            return new FileTestData(fileName, bytes);
        }
		public static void endConnection(Socket server){         
            if (!useHeader)
                endConnection1(server);
            else
                endConnection2(server);
		}
		private static void endConnection1(Socket server){
			String message = "end";
            byte[] byteMessage = Encoding.UTF8.GetBytes(message);
            int length = byteMessage.Length;
            byte[] b = BitConverter.GetBytes(length);
            Array.Reverse(b);
            server.Send(b);
            server.Send(byteMessage);

            server.Disconnect(true);
			server.Close();
		}

        private static void endConnection2(Socket server)
        {

            Header header = new Header(SocketMessageHelper.End, "", 0);
			server.Send(header.getByteHeader(), 55, SocketFlags.None);

            server.Disconnect(true);
            server.Close();
        }

		public static void saveFileToMemory(byte[] dataBytes, String fileName){
			var startTime = DateTime.Now;
			var documents = Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments);
			var filename = Path.Combine(documents, fileName);
			File.WriteAllBytes(filename, dataBytes);
			return;
		}

		public static FileTestData reachFromMemory(String fileName){
			var startTime = DateTime.Now;
            var documents = Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments);
            var filename = Path.Combine(documents, fileName);
			byte[] dataBytes = File.ReadAllBytes(filename);
			return new FileTestData(fileName, dataBytes);
		}
        
		public static byte[] generateFile(int fileSize){
            byte[] dataBytes = new byte[fileSize];
            new Random().NextBytes(dataBytes);
            return dataBytes;
		}

		public static void saveResult(int fileSize, double transferTime, String testName, String transferMedia, String action)
        {
            String results = "\n"
				    + DateTime.Now.ToString("R")
                    + ";" + action
				    + ";" + fileSize.ToString()
                    + ";" + transferTime.ToString()
                    + ";" + transferMedia
                    + ";" + testName + ";;;";

			String resultsFileName = "Results.csv";
            var documents = Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments);
			var filename = Path.Combine(documents, resultsFileName);
			File.AppendAllText(filename, results);
        }
      
    }
}

File: Memory_vs_sockets/src/com/iderek/ClientHelper.java
package com.iderek;

import java.io.*;
import java.net.Socket;
import java.net.SocketException;
import java.nio.ByteBuffer;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.*;

public class ClientHelper {
    int defaultBufferSize = 524288;
    SocketMessageHelper messageHelper;

    public ClientHelper(){
        messageHelper = new SocketMessageHelper();
    }

    public ArrayList<String> reachFileNamesFromServer(Socket server) throws IOException{
        InputStream in = server.getInputStream();
        OutputStream out = server.getOutputStream();

        String message = "viewing";
        byte[] byteMessage = message.getBytes();
        int length = byteMessage.length;
        byte[] b = ByteBuffer.allocate(4).putInt(length).array();
        out.write(b);
        out.write(byteMessage);

        b = new byte[4];
        in.read(b);
        int ok = ByteBuffer.wrap(b).getInt();

        b = new byte[4];
        in.read(b);
        int arrayLength = ByteBuffer.wrap(b).getInt();

        ArrayList<String> fileNames = new ArrayList<String>();
        for (int i = 0; i<arrayLength; i++){
            b = new byte[4];
            in.read(b);
            length = ByteBuffer.wrap(b).getInt();
            byteMessage = new byte[length];
            in.read(byteMessage);
            String name = new String(byteMessage, "UTF-8");
            fileNames.add(name);
        }

        return fileNames;
    }

    public ArrayList<String> reachFileNamesFromServer2(Socket server) throws IOException{
        InputStream in = server.getInputStream();
        OutputStream out = server.getOutputStream();

        Header header = new Header(SocketMessageHelper.viewing, "", 0);
        out.write(header.getByteHeader());

        byte[] b = new byte[4];
        in.read(b);
        int arrayLength = ByteBuffer.wrap(b).getInt();
        ArrayList<String> fileNames = new ArrayList<String>();
        for (int i = 0; i<arrayLength; i++){
            b = new byte[4];
            in.read(b);
            int length = ByteBuffer.wrap(b).getInt();
            byte[] byteMessage = new byte[length];
            in.read(byteMessage);
            String name = new String(byteMessage, "UTF-8");
            fileNames.add(name);
        }

        return fileNames;
    }

    public void sendFileToServer2(byte[] dataBytes,String fileName, Socket server, boolean serverCheck) throws IOException, InterruptedException{

        InputStream in = server.getInputStream();
        OutputStream out = server.getOutputStream();


        int length = dataBytes.length;

        Header header = new Header(SocketMessageHelper.sending, fileName, length);
        out.write(header.getByteHeader());


        ByteArrayInputStream byteStream = new ByteArrayInputStream(dataBytes);

        int bufferSize;
        if (length> defaultBufferSize)
            bufferSize = defaultBufferSize;
        else
            bufferSize = length;

        byte[]buffer = new byte[bufferSize];


        int count;
        while ((count = byteStream.read(buffer)) > 0 && server.isConnected()) {
            out.write(buffer, 0, count);
        }

        if (serverCheck) {
            byte[] b = new byte[1];
            in.read(b, 0, 1);
            if (b[0] != 1)
                throw new IOException("Server not ok!");
        }
        return;
    }

    public void sendFileToServer(byte[] dataBytes,String fileName, Socket server) throws IOException, InterruptedException{

        InputStream in = server.getInputStream();
        OutputStream out = server.getOutputStream();

        String message = "sending";
        byte[] byteMessage = message.getBytes();
        int length = byteMessage.length;
        byte[] b = ByteBuffer.allocate(4).putInt(length).array();
        out.write(b);
        out.write(byteMessage);

        b = new byte[4];
        in.read(b);
        int ok = ByteBuffer.wrap(b).getInt();

        message = fileName;
        byteMessage = message.getBytes();
        length = byteMessage.length;
        b = ByteBuffer.allocate(4).putInt(length).array();
        out.write(b);
        out.write(byteMessage);


        ByteArrayInputStream byteStream = new ByteArrayInputStream(dataBytes);
        length = dataBytes.length;

        int bufferSize;
        if (length> defaultBufferSize)
            bufferSize = defaultBufferSize;
        else
            bufferSize = length;

        b = ByteBuffer.allocate(4).putInt(length).array();
        out.write(b);

        byte[]buffer = new byte[bufferSize];

        b = new byte[4];
        in.read(b);
        ok = ByteBuffer.wrap(b).getInt();
        int count;
        while ((count = byteStream.read(buffer)) > 0 && server.isConnected()) {
            out.write(buffer, 0, count);
        }

        b = new byte[4];
        in.read(b);
        length = ByteBuffer.wrap(b).getInt();
        byteMessage = new byte[length];
        in.read(byteMessage);
        String serverMessage = new String(byteMessage, "UTF-8");
        if (!serverMessage.equals("recieved"))
            throw new SocketException(serverMessage);
        System.out.println("Server: " +  serverMessage);

        return;
    }

    public class FileTestData{
        public String name;
        public byte[] data;

        public FileTestData(){}

        public FileTestData(String name, byte[] data){
            this.name = name;
            this.data = data;
        }
    }

    public FileTestData reachFileFromServer2(Socket server, String fileName) throws IOException, InterruptedException{
        InputStream in = server.getInputStream();
        OutputStream out = server.getOutputStream();

        Header header = new Header(SocketMessageHelper.requesting, fileName, 0);
        out.write(header.getByteHeader());

        byte []b = new byte[4];
        in.read(b, 0,4);
        int length = ByteBuffer.wrap(b).getInt();

        int bufferSize;
        if (length> defaultBufferSize)
            bufferSize = defaultBufferSize;
        else
            bufferSize = length;


        byte[] buffer = new byte[bufferSize];
        byte[] bytes;
        ByteBuffer dataBuffer = ByteBuffer.allocate(length + length);
        int n = length;
        int dataRead;
        while ((dataRead = in.read(buffer, 0, bufferSize)) != -1) {
            n -= dataRead;
            dataBuffer.put(buffer,0, dataRead);
            if (n <= 0)
                break;
            if (n <= bufferSize)
                bufferSize = n;
        }

        bytes = Arrays.copyOfRange(dataBuffer.array(), 0,length);

        dataBuffer.clear();

        return new FileTestData(fileName, bytes);
    }

    public FileTestData reachFileFromServer(Socket server, String fileName) throws IOException, InterruptedException{
        InputStream in = server.getInputStream();
        OutputStream out = server.getOutputStream();

        String message = "requesting";
        byte []byteMessage = message.getBytes();
        int length = byteMessage.length;
        byte[] b = ByteBuffer.allocate(4).putInt(length).array();
        out.write(b);
        out.write(byteMessage);

        b = new byte[4];
        in.read(b);
        int ok = ByteBuffer.wrap(b).getInt();

        byteMessage = fileName.getBytes();
        length = byteMessage.length;
        b = ByteBuffer.allocate(4).putInt(length).array();
        out.write(b);
        out.write(byteMessage);

        b = new byte[4];
        in.read(b);
        length = ByteBuffer.wrap(b).getInt();
        byteMessage = new byte[length];
        in.read(byteMessage);
        String serverMessage = new String(byteMessage, "UTF-8");
        System.out.println("Server: " +  serverMessage);

        b = new byte[4];
        in.read(b);
        length = ByteBuffer.wrap(b).getInt();

        int bufferSize;
        if (length> defaultBufferSize)
            bufferSize = defaultBufferSize;
        else
            bufferSize = length;

        b = new byte[4];
        in.read(b);
        ok = ByteBuffer.wrap(b).getInt();
        Thread.sleep(1);
        byte[] buffer = new byte[bufferSize];
        byte[] bytes = new byte[length];
        ByteBuffer dataBuffer = ByteBuffer.allocate(length + length);
        String fromBytes = new String();
        int n = length;
        int dataRead;
        while ((dataRead = in.read(buffer,0, bufferSize)) != -1) {
            n -= dataRead;
            dataBuffer.put(buffer,0, dataRead);
            if (n <= 0)
                break;
            if (n <= bufferSize)
                bufferSize = n;
        }

        bytes = Arrays.copyOfRange(dataBuffer.array(), 0,length);

        dataBuffer.clear();

        message = "recieved";
        byteMessage = message.getBytes();
        length = byteMessage.length;
        b = ByteBuffer.allocate(4).putInt(length).array();
        out.write(b);
        out.write(byteMessage);

        return new FileTestData(fileName, bytes);
    }

    public void saveFileToMemory(byte[] dataBytes,String fileName, String path) throws IOException{
        Path file = Paths.get(path+"/"+fileName);
        try{
            Files.write(file, dataBytes);
            return;
        }catch (IOException e){
            System.out.println(e.toString());
        }catch ( NullPointerException a){
            System.out.println("data not initialised!");
        }
        return ;
    }

    public FileTestData reachFromMemory(String fileName, String path){

        Path nFile = Paths.get(path+"/"+fileName);
        try{
            byte[] dataBytes = Files.readAllBytes(nFile);
            return new FileTestData(fileName, dataBytes);
        }catch (IOException e){
            System.out.println(e.toString());
            e.printStackTrace();
        }catch ( NullPointerException a){
            System.out.println("data not initialised!");
            a.printStackTrace();
        }
        return null;
    }

    public void saveResult(int fileSize, double transferTime, String testName, String transferMedia, String action){
        String results = "\n"
                + new Date(System.currentTimeMillis())
                + ";" + action
                + ";" + fileSize
                + ";" + String.format("%.3f",transferTime)
                + ";" + transferMedia
                + ";" + testName+";;;";

        try{
            String resultsFile= "Results.csv";
            FileWriter fw = new FileWriter(resultsFile,true);
            fw.write(results);
            fw.close();
        }
        catch(IOException ioe){
            System.out.println("IOException: " + ioe.getMessage());
        }
    }

    public byte[] generateFile(int fileSize){
        byte[] dataBytes = new byte[fileSize];
        new Random().nextBytes(dataBytes);
        return dataBytes;
    }
}
